plugin/Utils.pm

```diff
@@ -172,13 +201,65 @@ sub getPlaylist {
         $errorCb->("Error getting play list ($error)");
       }
     },
-      stationToken => $stationToken
+      stationToken => $stationId
     );
   };
 
-  getStationToken($stationId, $withStationToken, $onError);
+  getWebService($withWebsvc, $onError);
 }
 
+sub _rest {
+  my ($endpoint, $args, $successCb, $errorCb, $passthrough) = @_;
 
-1;
+  # Fetch CSRF cookie from www.pandora.com
+  my $uri = URI->new('https://www.pandora.com');
+  my $http = Slim::Networking::SimpleAsyncHTTP->new(
+    sub {
+      _restCsrfCallback->($endpoint, $args, $successCb, $errorCb, $passthrough);
+    },
+    sub {
+      my ($req, $error) = @_;
+      $log->error($error);
+      $errorCb->($error, $passthrough);
+    },
+    {
+      $args,
+    }
+  );
+  $http->head($uri);
+}
+
+sub _restCsrfCallback {
+  my ($endpoint, $args, $successCb, $errorCb, $passthrough) = @_;
+  my $csrftoken = Slim::Networking::Async::HTTP->cookie_jar->get_cookies('.pandora.com', 'csrftoken');
+  unless ($csrftoken) {
+    $errorCb->(result => undef, error => 'Failed to retrieve CSRF token from www.pandora.com');
+    return;
+  }
+  my $uri = URI->new("https://www.pandora.com/api/${endpoint}");
+  my $withWebsvc = sub {
+    my ($websvc) = @_;
+    my $http = Slim::Networking::SimpleAsyncHTTP->new(
+      sub {
+        my $response = shift;
+        my $json = JSON->new->decode($response->content);
+        $successCb->(result => $json, passthrough => $passthrough);
+      },
+      sub {
+        my ($req, $error) = @_;
+        $log->error($error);
+        $errorCb->($error, $passthrough);
+      }
+    );
+    $http->post(
+      $uri,
+      'X-CsrfToken' => $csrftoken,
+      'X-AuthToken' => $websvc->{'userAuthToken'},
+      'Content-Type' => 'application/json;charset=utf-8',
+      JSON->new->encode($args),
```

Comment:
Might be good to encode with UTF-8 here?  Probably doesn't affect getting the station list, but might be needed if Pyrrha ever uses any of the other REST APIs?

Reply:
Fixed in https://github.com/sspiff/lms-plugin-pyrrha/pull/20/commits/3ca9c4d9502013bd229857016651f8a458468f13